### DISCUSSION TITLE
Developer Preview of upcoming Release 2.4.0

### DISCUSSION THREAD

**Discussion by mhhen** [aws/aws-cdi-sdk#73] [Announcements]
We have made the developer preview version of the upcoming Release 2.4.0 available [here](https://github.com/aws/aws-cdi-sdk/tree/988ea55f51d8d16d8f36d326ba0bdbb7f1e7301d), on the "Developer Preview" branch.

**Comment by f2bo** [comment]
Probably caused by new compiler warnings, but I'm not getting a clean build in this branch with gcc 10.3.0, so it fails with `-Werr`. For example:
```
/workspace/aws-cdi-sdk/src/test/test_console.c: In function 'vprintf_line':
/workspace/aws-cdi-sdk/src/test/test_console.c:155:13: error: format not a string literal and no format arguments [-Werror=format-security]
  155 |             printf(prefix_str);
      |             ^~~~~~
cc1: all warnings being treated as errors
make: *** [Makefile:409: /workspace/aws-cdi-sdk/build/debug/obj/test_console.o] Error 1
```
and
```
/workspace/aws-cdi-sdk/src/cdi/adapter_efa_probe_rx.c: In function 'ProbeRxControlProcessProbeState':
/workspace/aws-cdi-sdk/src/cdi/adapter_efa_probe_rx.c:390:48: error: implicit conversion from 'enum <anonymous>' to 'ProbeState' [-Werror=enum-conversion]
  390 |             probe_ptr->rx_probe_state.rx_state = kProbeCommandReset; // Advance to resetting state.
      |                                                ^
cc1: all warnings being treated as errors
make: *** [Makefile:399: /workspace/aws-cdi-sdk/build/debug/obj/adapter_efa_probe_rx.o] Error 1
```
There are also some Doxygen errors with v1.9.1.
```
warning: Tag 'COLS_IN_ALPHA_INDEX' at line 973 of file 'doc/Doxyfile' has become obsolete.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u"
/workspace/aws-cdi-sdk/src/cdi/private.h:160: warning: argument 'Pointer' of command @param is not found in the argument list of CdiCallback(const void *param_ptr)
/workspace/aws-cdi-sdk/src/cdi/private.h:162: warning: The following parameter of CdiCallback(const void *param_ptr) is not documented:
  parameter 'param_ptr'
make: *** [Makefile:455: docs] Error 1
```

**Comment by mhhen** [comment]
Thanks for helping identify some compile issues with GCC 10.3.0. We resolved them and pushed changes to the developer_preview branch. See changes [here](https://github.com/aws/aws-cdi-sdk/commit/7954d918c57ed6ce4a8271416b47425919853e75).

**Reply by f2bo** [reply]
Thanks for fixing these issues. Keep in mind that there were some Doxygen errors in `src/cdi/private.h` too. 

There also seems to be a regression in the **cdi_test** app. In this branch, the receiver does not update its statistics, which always remain as zero, and does not exit when finished. The transmitter does update its stats.

![image](https://user-images.githubusercontent.com/1148767/161796975-58d0fc1b-1b0b-45e6-afb8-1c844175784f.png)

**Reply by mhhen** [reply]
Having trouble trying to reproduce the problem with **cdi_test** you noted above. Not sure how to proceed other than asking for the cdi_test command lines you used. Will be updating the developer_preview with the doxygen errors fixed and a few other changes.

**Reply by f2bo** [reply]
Strange. I see the zero stats problem both in Windows and Linux with the `development_preview` branch. If I switch back to `mainline` the problem disappears. 

For example:

## Receiver
```
cdi_test --adapter SOCKET_LIBFABRIC --local_ip 127.0.0.1 -X --rx RAW --dest_port 2000 --rate 24 --num_transactions 300 -S --payload_size 5184000 --pattern INC --log_level DEBUG --multiwindow
```
## Transmitter
```
cdi_test --adapter SOCKET_LIBFABRIC --local_ip 127.0.0.1 -X --tx RAW --remote_ip 127.0.0.1 --dest_port 2000 --rate 24 --num_transactions 300 -S --payload_size 5184000 --tx_timeout 100000 --keep_alive --multiwindow
```
Also, in Linux, they don't exit correctly after transmitting all the payloads and I need to kill them. Moreover, in Windows, the transmitter crashes freeing memory when exiting in the `TestConsoleDestroy` function. 
```
---------------------------
Microsoft Visual C++ Runtime Library
---------------------------
Debug Error!

Program: F:\Projects\aws\aws-cdi-sdk\proj\x64\Debug\cdi_test.exe

Damage before 0x0000019D4A635520 which was allocated by aligned routine
```
and the call stack is:

```
ucrtbased.dll!_aligned_free_dbg(void * block) Line 1978	C++
ucrtbased.dll!_aligned_free(void * block) Line 545	C++
cdi_test.exe!CdiOsMemFree(void * mem_ptr) Line 823	C
cdi_test.exe!TestConsoleDestroy(bool _abnormal_termination) Line 336	C
cdi_test.exe!FreeAppLoggerResources() Line 232	C
cdi_test.exe!main(int argc, const char * * argv) Line 491	C
```

**Reply by mhhen** [reply]
Thanks for providing the cdi_test command lines. It turns out the problem is caused by the large payload size (5,184,000 bytes) when using the libfabric socket adapter. This problem will not occur when using the EFA adapter. We now use limit the requested number of Rx adapter buffers, based on information provided by libfabric. In the socket adapter case, the limit is 385 buffers, far less than the amount required for the large payload. To resolve, we will push a solution for this and your other recent issues to the developer_preview branch and will update this discussion.

**Reply by f2bo** [reply]
Good. So this is caused by a change in behavior in this branch, correct? I've been using that payload size and the libfabric adapter forever.  

I had noticed the following message when I copied the output from the **cdi_test_min_rx** app in this issue https://github.com/aws/aws-cdi-sdk/issues/67#issuecomment-1087818509 yesterday, which was new to me. I meant to ask, but since it was only a warning and it didn't seem relevant to that issue, I decided not to. Is this related?

```
[2022-04-04T12:23:10-03:00] WARNING: [CreatePacketPool:175] Requested Rx packet buffer count[12000] exceeds endpoint capability[376]. Reducing.
```

I've tested again with smaller payload sizes, both in Windows and Linux, and the stats now update correctly and I no longer see the crash in Windows. I notice that the "Requested Rx packet buffer count exceeds endpoint capability. Reducing." even for small payload sizes. Is this expected?

And this was a quick test, but in Linux I still saw a few cases where the apps did not shut down correctly and I had to kill them, even though the receiver stats did update correctly. I'll test again once you update the branch. 

Thanks!

**Reply by f2bo** [reply]
After further testing, I started seeing the crash in Windows again and realized that it's unrelated to the payload size but rather to the use of `--multiwindow`. 

It seems to be caused by calling `FreeAppLoggerResources` twice in `cdi_test.c`, first in line 474 and then again in line 490, when shutting down the application.

https://github.com/aws/aws-cdi-sdk/blob/7954d918c57ed6ce4a8271416b47425919853e75/src/test/cdi_test.c#L472-L491

**Reply by mhhen** [reply]
Thank you for providing additional detail. We pushed a new revision to the developer_preview branch that should resolve the issues described above.

**Reply by mhhen** [reply]
Were you able to try out the new developer_preview version to see if it resolved your issues?

**Reply by f2bo** [reply]
Yes. I wanted to do some more testing before replying. 

The crash in Windows when using `--multiwindow` seems to be resolved. However, I still see the problem with large payload sizes using the libfabric adapter. I usually run the Linux version inside a Docker container and with these latest changes, it also seems to destabilize Docker itself whenever the app is executed with payloads sizes above 3.25MB, which becomes unresponsive until I manage to open another terminal and kill the receiver. This wasn't the case before these changes.

Just to be clear, If this works correctly with EFA, I can live with the smaller payload sizes for development.

**Reply by mhhen** [reply]
Yes, that works correctly with EFA. See developer_preview release note [here](https://github.com/aws/aws-cdi-sdk/blob/38ebe5e60380e07f02b649d7487b5baaae78645f/RELEASE_NOTES#L25) about Rx buffer and payload sizes.
